src/ConnectedMode.UnitTests/Suppressions/SuppressionIssueStoreUpdaterTests.cs

```diff
@@ -0,0 +1,295 @@
+                .Callback(() => callSequence.Add("SwitchToBackgroundThread"));
+
+            var testSubject = CreateTestSubject(queryInfo.Object,
+                threadHandling: threadHandling.Object);
+
+            await testSubject.UpdateAllServerSuppressionsAsync();
+
+            queryInfo.Invocations.Should().HaveCount(1);
+            threadHandling.Invocations.Should().HaveCount(1);
+
+            callSequence.Should().ContainInOrder("SwitchToBackgroundThread", "GetProjectKeyAndBranchAsync");
+        }
+
+        [TestMethod]
+        public void UpdateAll_CriticalExpression_NotHandled()
+        {
+            var queryInfo = new Mock<IServerQueryInfoProvider>();
+            queryInfo.Setup(x => x.GetProjectKeyAndBranchAsync(It.IsAny<CancellationToken>()))
+                .Throws(new StackOverflowException("thrown in a test"));
+
+            var logger = new TestLogger(logToConsole: true);
+
+            var testSubject = CreateTestSubject(queryInfo.Object, logger: logger);
+
+            Func<Task> operation = testSubject.UpdateAllServerSuppressionsAsync;
+            operation.Should().Throw<StackOverflowException>().And.Message.Should().Be("thrown in a test");
+
+            logger.AssertPartialOutputStringDoesNotExist("thrown in a test");
+        }
+
+        [TestMethod]
+        public void UpdateAll_NonCriticalExpression_IsSuppressed()
+        {
+            var queryInfo = new Mock<IServerQueryInfoProvider>();
+            queryInfo.Setup(x => x.GetProjectKeyAndBranchAsync(It.IsAny<CancellationToken>()))
+                .Throws(new InvalidOperationException("thrown in a test"));
+
+            var logger = new TestLogger(logToConsole: true);
+
+            var testSubject = CreateTestSubject(queryInfo.Object, logger: logger);
+
+            Func<Task> operation = testSubject.UpdateAllServerSuppressionsAsync;
+            operation.Should().NotThrow<InvalidOperationException>();
+
+            logger.AssertPartialOutputStringExists("thrown in a test");
+        }
+
+        [TestMethod]
+        public void UpdateAll_OperationCancelledException_CancellationMessageLogged()
+        {
+            var queryInfo = new Mock<IServerQueryInfoProvider>();
+            queryInfo.Setup(x => x.GetProjectKeyAndBranchAsync(It.IsAny<CancellationToken>()))
+                .Throws(new OperationCanceledException("thrown in a test"));
+
+            var logger = new TestLogger(logToConsole: true);
+
+            var testSubject = CreateTestSubject(queryInfo.Object, logger: logger);
+
+            Func<Task> operation = testSubject.UpdateAllServerSuppressionsAsync;
+            operation.Should().NotThrow<OperationCanceledException>();
+
+            logger.AssertPartialOutputStringDoesNotExist("thrown in a test");
+            logger.AssertOutputStringExists(Resources.Suppressions_FetchOperationCancelled);
+        }
+
+        [TestMethod]
+        public async Task UpdateAll_CallInProgress_CallIsCancelled()
```

Comment:
This test is complicated, but it's the only way to force the operations to occur in the expected order i.e. 

1. first call starts
2. second call starts and cancels first call
3. first call notices it has been cancelled and aborts
4. second call finishes.

To make sure the operations happen in a specific order we need to use `ManualResetEvent`s to block the execution at various points.

Reply:
Ok yes that is a long test but the way its written and commented on its understandable so Its okay